File: Connector/CSharp/Session.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace Sv
{
    public class Session
    {
        private Socket _socket;
        private byte[] _buffer = new byte[1024];
        int _isDisconnected;
        private SocketAsyncEventArgs _recvEvent = new SocketAsyncEventArgs();

        public Session(Socket socket)
        {
            _socket = socket;
            _isDisconnected = 0;
        }

        public void Run()
        { 
            _recvEvent.Completed += (object sender, SocketAsyncEventArgs recvEvent) =>
            {
                if (recvEvent.SocketError == SocketError.Success)
                {
                    OnReceive(new Span<byte>(_recvEvent.Buffer, 0, _recvEvent.Buffer.Length));
                    _socket.ReceiveAsync(_recvEvent);
                }
                else
                {
                    OnDisconnected();
                }
            };
            _recvEvent.SetBuffer(_buffer);

            _socket.ReceiveAsync(_recvEvent);
        }

        public void Send(Span<byte> data)
        {
            _socket.Send(data);
        }

        public void Disconnect()
        {
            if (Interlocked.Exchange(ref _isDisconnected, 1) == 0)
            {
                _socket.Disconnect(false);
                OnDisconnected();
            }
        }

        public virtual void OnConnected() { }
        public virtual void OnDisconnected() { }
        public virtual void OnReceive(Span<byte> data) { }
    }
}

File: SvEngine/src/Network/Session.cpp
//
// Created by han93 on 2023-12-13.
//
#include "pch.h"

#include <Network/Session.hpp>

#include "net/Context.hpp"
#include "net/Exception.hpp"

CREATE_ARENA(Session, 0x1000)

Session::Session() : m_buffer(0x10000, '\0'), m_isDisconnected(false), m_flushSend(false) {
}

void Session::Run(std::shared_ptr<Socket> sock) {
    m_sock = sock;

    m_recvCtx.completed = bind(&Session::OnRecvCompleted, this, std::placeholders::_1, std::placeholders::_2);
    m_recvCtx.buffer = m_buffer;
    m_sendCtx.completed = bind(&Session::OnSendCompleted, this, std::placeholders::_1, std::placeholders::_2);

    m_sock->receive(&m_recvCtx);
    m_ref = shared_from_this();
}

void Session::OnRecvCompleted(Context *context, bool isSuccess) {
    if(!isSuccess || context->length == 0) {
        Disconnect();
        return;
    }
    OnReceive(context->buffer.subspan(0, context->length), context->length);
    m_sock->receive(context);
}

void Session::OnSendCompleted(Context* context, bool isSuccess)
{
    m_sendCtx.sendBuffer.clear();
    m_flushSend.store(false);
}

Session::~Session() {
}

void Session::Disconnect() {
    if (!m_isDisconnected.exchange(true))
    {
        OnDisconnected(m_sock->getRemoteEndpoint().value());
        m_ref = nullptr;
    }
}

Socket Session::GetSocket() {
    return *m_sock;
}

void Session::Send(std::span<char> buffer)
{
    m_sendCtx.sendBuffer.push_back(buffer);
    if (!m_flushSend.exchange(true))
    {
        m_sock->send(&m_sendCtx);
    }
}

void Session::Send(Packet* packet) {
    packet->Write();
    Send(packet->Data());
}
